### DISCUSSION TITLE
--west-flash Causes Twister Failure?

### DISCUSSION THREAD

**Discussion by Maxxus220** [zephyrproject-rtos/zephyr#69204] [Q&A]
Hi, I was running into some issues trying to get twister to run for some testing. I wanted to check if I'm doing something wrong before raising an issue for it. 

When running the following on Windows 11 with multiple devices connected to JLinks (need to specify which one to run on):
```
python -m west twister --short-build-path --device-testing --device-serial COM15 --west-flash"-i 1050253837" -p nrf52840dk_nrf52840 -T .\zephyr\tests\drivers\counter\counter_nrf_rtc\fixed_top -vv
```

I get the following error:
```
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) tests\drivers\counter\counter_nrf_rtc\fixed_top\drivers.counter.nrf_rtc on nrf52840dk_nrf52840 error (Device issue (Flash error?))
```

If I build and flash the test myself, it works just fine and I can see the output on COM15:
```
python -m west build -b nrf52840dk_nrf52840 .\zephyr\tests\drivers\counter\counter_nrf_rtc\fixed_top
python -m west flash -i 1050253837
```

```
*** Booting Zephyr OS build zephyr-v3.5.0 ***
Running TESTSUITE counter
===================================================================
START - test_set_custom_top_value_fails
 PASS - test_set_custom_top_value_fails in 0.201 seconds
===================================================================
START - test_top_handler
 PASS - test_top_handler in 0.221 seconds
===================================================================
TESTSUITE counter succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [counter]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.422 seconds
 - PASS - [counter.test_set_custom_top_value_fails] duration = 0.201 seconds
 - PASS - [counter.test_top_handler] duration = 0.221 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

It seems this issue has something to do with using the --west-flash flag. When I disconnect the other JLink and run the same command without the --west-flash flag, twister passes without issues. Even without the other JLink connected, introducing the --west-flash flag with or without ="-i <serial_number>" causes twister to fail with (Device issue (Flash error?)). If both JLinks are connected, running without the --west-flash flag doesn't work as I need to specify which device is being tested.

Has anyone else run into this or is there something I'm doing wrong? I've attached the output of twister --generate-hardware-map and the twister.log, device.log, build.log, and handler.log for:
```
python -m west twister --short-build-path --device-testing --device-serial COM15 --west-flash -p nrf52840dk_nrf52840 -T .\zephyr\tests\drivers\counter\counter_nrf_rtc\fixed_top\ -vv
```
With only one JLink connected.

[twister.log](https://github.com/zephyrproject-rtos/zephyr/files/14334306/twister.log)
[build.log](https://github.com/zephyrproject-rtos/zephyr/files/14334308/build.log)
[device.log](https://github.com/zephyrproject-rtos/zephyr/files/14334309/device.log)
[handler.log](https://github.com/zephyrproject-rtos/zephyr/files/14334310/handler.log)
[hwmap.txt](https://github.com/zephyrproject-rtos/zephyr/files/14334313/hwmap.txt)

**Comment by prayassamriya** [comment]
@Maxxus220  we ran into similar issue. we have GitHub actions to test twister. each physical agent(windows machine) is connected to multiple boards.
Did you find the fix for your problem?

**Reply by Maxxus220** [reply]
@prayassamriya It's been a long time since I looked at what I did, but I think the solution I found was using the `--hardware-map` option instead. You pass a hardware map file. Mine looks something like this:
```
- connected: true
              id: your_probe_id
              platform: nrf52840dk_nrf52840
              product: J-Link
              runner: nrfjprog
              serial: your_com_port
```

There are also examples of the hardware map file here: https://docs.zephyrproject.org/latest/develop/test/twister.html
They can also be generated. In my case, I set up the map file to only have one board so that it was forced to use that one.

**Reply by prayassamriya** [reply]
Okay, thanks for information.
